### PR TITLE
Feat/submission save current answer

### LINKED ISSUE #813: Save "changes" when loading a former submission
## Overview

When clicking "load" on a submission, save the current state/answer if it has changed. 

## Discussion

* It probably requires to be able to track if a task has been changed since its last loading (from a specific load) or from the task "empty" state (the state when first loaded).
* Use [answerCreate](https://france-ioi.github.io/algorea-devdoc/api/#operation/answerCreate) to save an answer

## Acceptance criteria
- [ ] When opening a task for the first time, if I load a previous submission without having changed anything, there is no save.
- [ ] After having loaded a former submission, if I load another submission without having changed anything, there is no save.
- [ ] If I load a former submission, the (latest, not just the latest saved current-answer) change is automatically saved
- [ ] If I do a change and immediately load a former submission (while the auto-save of the current state has not been done yet), the (latest!) change is automatically saved
- [ ] Make sure the saved answers are listed in the progress list and can be loaded as well

### PR DESCRIPTION
**Pull request by ThomasJuster**
## Description

Fixes #813 

## Notes (out of scope, known issues, hints for reviewing code, ...)  (optional)

After some discussions, the issue has changed a bit:
- Saving the current-answer-to-override with former answer happens at task load rather than at "Load" link click
- Because of that, acceptance criteria indicating "no save" cannot be implemented. Concerned use cases:
  - `When opening a task for the first time, if I load a previous submission without having changed anything, there is no save.`
  - `After having loaded a former submission, if I load another submission without having changed anything, there is no save.`
- [This other PR](https://github.com/France-ioi/AlgoreaFrontend/pull/818) implicitly covers 2 use cases:
  - `If I load a former submission, the (latest, not just the latest saved current-answer) change is automatically saved`
  - `If I do a change and immediately load a former submission (while the auto-save of the current state has not been done yet), the (latest!) change is automatically saved`
- As for the last acceptance criteria `Make sure the saved answers are listed in the progress list and can be loaded as well`, it relies on [this backend PR](https://github.com/France-ioi/AlgoreaBackend/pull/808)

&nbsp;

To sum up, none of the acceptance criteria are handled by this PR, what this PR covers is:
- When loading the former answer, remove in url the path parameter `answerId` to avoid reloading the former answer when refreshing the page
- When initializing the task with a former answer id,
  - fetch current answer then save it as "Saved"
  - fetch former answer then save it as "Current" (only once former current is saved as "Saved")
- Handle "Saved" and "Current" answers in activity log service, allow to load "Saved" answers

## Test cases

- [ ] Case 1: `answerId` removal from url
  1. Given I am the usual user
  2. When I go to [this task activity/progress](https://dev.algorea.org/branch/feat/submission-save-current-answer/en/#/activities/by-id/784078458149384669;path=4702,1352246428241737349,314613032161178344,34689573454313988;parentAttempId=0/details/progress/history)
  3. And I load any former submission
  4. Then I see the page is loaded with an `answerId` as path parameter
  5. And I wait a bit
  6. Then I see the `answerId` path parameter has been removed

- [ ] Case 2: initializing task with former answer id
  1. Given I am a brand new user
  2. When I go to [this task](https://dev.algorea.org/branch/feat/submission-save-current-answer/en/#/activities/by-id/784078458149384669;path=4702,1352246428241737349,314613032161178344,34689573454313988;parentAttempId=0/details)
  3. And I solve first level
  4. And I add some blocks on second level
  5. And I click on "Progress tab"
  6. (optional) And I open developer tools on network panel (throttle network to clearly see it in action)
  7. And I load the (only) former answer
  8. Then I should be redirected to the task
  9. Then 2nd level should now be empty because task should have saved former current answer (containing 2nd level changes) and load former answer as new current answer
  10. And I navigate back to progress
  11. Then I should see two answers, one "Saved" and one "Submission"
  12. And I load the saved one
  13. Then I should have the same blocks I added before loading former answer (at step 4)

## No regression tests

- [ ] Case 1:
  1. Given I am the usual user
  2. When I go to [this task with bad answer id in url](https://dev.algorea.org/branch/feat/submission-save-current-answer/en/#/activities/by-id/784078458149384669;path=4702,1352246428241737349,314613032161178344,34689573454313988;parentAttempId=0;answerId=1000/details)
  3. Then I should see "Unable to load task, [link] load your most recent answer [/link]`
  4. And I click on link
  5. Then the task should be reloaded with current answer

### CODE REVIEW

**Comment by smadbe** [comment]
Backend redeployed with the log service return all kind of submissions.

File: src/app/modules/item/pages/item-by-id/item-by-id.component.ts [hunk 1]
```diff
@@ -55,7 +55,8 @@ export class ItemByIdComponent implements OnDestroy {
 
     // on route change or user change: refetch item if needed
     this.activatedRoute.paramMap.pipe(
-      repeatLatestWhen(this.userSessionService.userChanged$)
+      repeatLatestWhen(this.userSessionService.userChanged$),
+      filter(() => !(history.state as Record<string, boolean>).preventRefetch),
```
Comment (smadbe) [inline src/app/modules/item/pages/item-by-id/item-by-id.component.ts:59]: a quick comment may be useful to remember what it is for

**Review by smadbe — APPROVED** [review]
👍

File: src/app/modules/item/pages/item-by-id/item-by-id.component.ts (hunk 1 above)
Comment (ThomasJuster) [reply src/app/modules/item/pages/item-by-id/item-by-id.component.ts:62]: fixed

**Pull request merged**